python/ray/_raylet.pyx

```diff
@@ -551,6 +559,15 @@ cdef void async_plasma_callback(CObjectID object_id,
             event_handler._loop.call_soon_threadsafe(
                 event_handler._complete_future, obj_id)
 
+cdef void kill_main_task() nogil:
+    with gil:
+        # This prevents this interrupt from being called to early. It may still
+        # be called later, but will not cancel another task because this is
+        # called with a C++ mutex lock.
```

Comment:
We might be able to address this by always calling `PyErr_CheckSignals()` in some known location (e.g., just before returning from cython to c++), but it'll take some thought. The python interpreter doesn't give a guarantee on when this will be raised unless you call it manually.